aspnetcore/razor-components/components.md

```diff
@@ -761,3 +761,54 @@ The time is 10/04/2018 01:26:52.
 
 Your pet's name is Rex.
 ```
+
+## Dynamic components
+
+Dynamic components can be used to dynamically generate content. `Microsoft.AspNetCore.Components.RenderTree` provides methods for manipulating components and elements.
+
+> [!WARNING]
+> Dynamic components are an advanced scenario and aren't recommended for general use. Dynamically generated components can bypass tag structure safety checks. Inadvertently rendering a faulty tag in the browser can result in a security vulnerability.
+>
+> When creating components dynamically, **always** hardcode the sequence number. Do **not** generate the sequence number using a calculation or counter.
```

Comment:
What is the motivation for this guidance?

Reply:
Same thing: Steve said to call this out ... he made it sound rather important. He implied that if the sequence number is calculated or based on a counter that it can result in malformed content.

Reply:
The reason is that the diff algorithm relies on the sequence numbers corresponding to distinct lines of code, not distinct call invocations. If developers get this wrong, it won’t lead to anything being visibly  malformed, but will mean the diff produces a suboptimal output (as in, perf is worse). 

It’s problematic precisely because developers don’t realise they have caused a perf problem by trying to be too smart.

Reply:
Thx ... I'll patch it up now.

Reply:
Given that the logic for dynamically generated rendering code may be arbitrarily complicated it's not clear to me that the sequence numbers can always be static and hard coded. For example, what if I'm adding N components where N is read from a database? It feels like there is a constraint on sequence numbers here that is not being clearly articulated. @SteveSandersonMS?

Reply:
Sequence numbers are source code line numbers. Never generate them dynamically.

In your example, the N is irrelevant. It would never be an aspect of sequence numbers.

Reply:
So my for loop should look like this (where the sequence numbers never change)?:

```
int n = ReadCountFromDb();
for (int i = 0; i < n; i++) {
        builder.OpenComponent(0, typeof(PetDetails));
        builder.AddAttribute(1, "PetDetails", "Someone's best friend");              
        builder.CloseComponent();
}
```

Reply:
Exactly!

Reply:
OK - I think this deserves more than just a note. We should add some text to this section that describes what the sequence numbers are and how they are used to motivate any guidance on how to specify them correctly.

Reply:
I'll draft some text (and move it out of the NOTE into the text) on the next commit within the hour.